DaemonSet updates - take 2

lukaszo opened the pull request:
It implements https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md
Feature kubernetes/features#124
#15310

```release-note
Implement the update feature for DaemonSet.
```

cc @kargakis @janetkuo @mikedanese

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41116)
<!-- Reviewable:end -->

kargakis commented:
@kubernetes/sig-apps-pr-reviews

kargakis commented on pkg/apis/extensions/types.go:
```diff
@@ -482,6 +468,20 @@ type DaemonSetStatus struct {
+	// available (ready for at least minReadySeconds)
+	NumberAvailable int32
+
+	// NumberUnavailable is the number of nodes that should be running the
```
FWIW, for Deployments, this is the number of pods running but not considered ready/available.
https://github.com/kubernetes/kubernetes/blob/d953402cdf414a877cb537d8db46646db38e7bef/pkg/controller/deployment/sync.go#L586

janetkuo replied:
That means the same thing as `!(running && available)` as this comment described, right?

kargakis commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -219,6 +225,73 @@ func Convert_v1beta1_RollingUpdateDeployment_To_extensions_RollingUpdateDeployme
 	return nil
 }
 
+func Convert_extensions_DaemonSetSpec_To_v1beta1_DaemonSetSpec(in *extensions.DaemonSetSpec, out *DaemonSetSpec, s conversion.Scope) error {
```
Why do we need custom conversions?

lukaszo replied:
Hmm, I had some errors with validation earlier and thought that it will help. It turns out that only
`Convert_extensions_RollingUpdateDaemonSet_To_v1beta1_RollingUpdateDaemonSet                 Convert_v1beta1_RollingUpdateDaemonSet_To_extensions_RollingUpdateDaemonSet,`
are required

kargakis commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -141,6 +144,38 @@ func ValidateDaemonSetSpec(spec *extensions.DaemonSetSpec, fldPath *field.Path)
+		}
+		allErrs = append(allErrs, ValidateRollingUpdateDaemonSet(strategy.RollingUpdate, fldPath.Child("rollingUpdate"))...)
+	} else {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("type"), strategy.Type, "RollingUpdate and OnDelete are only supported types"))
```
use field.NotSupported

lukaszo replied:
done

kargakis commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -141,6 +144,38 @@ func ValidateDaemonSetSpec(spec *extensions.DaemonSetSpec, fldPath *field.Path)
+
+func ValidateDaemonSetUpdateStrategy(strategy *extensions.DaemonSetUpdateStrategy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if strategy.Type == extensions.OnDeleteDaemonSetStrategyType {
```
nit: make this a switch

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -548,10 +560,12 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
 	glog.V(4).Infof("Nodes needing daemon pods for daemon set %s: %+v, creating %d", ds.Name, nodesNeedingDaemonPods, createDiff)
 	createWait := sync.WaitGroup{}
 	createWait.Add(createDiff)
+	dsPodTemplateSpecHash := podutil.GetPodTemplateSpecHashFnv(ds.Spec.Template)
```
Add a TODO to this helper to be moved in a more generic location.

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,75 @@
+			return true
+		}
+	} else {
+		// XXX: Hash does not exist. It's old pod. For now returning true.
```
In such a case I think you should compare (apimachinery.Semantic.DeepEqual) the ds pod template spec with the pod spec AND the labels and annotations found in the metadata of the ds podtemplate vs pod metadata.

janetkuo replied:
From @lukaszo's earlier comment it seems hard to compare pod spec with DS template because a lot of default values are applied while a pod is created. @lukaszo does this mean we won't be able to rolling update a DS created prior to 1.6?

kargakis replied:
Why can't we have a helper that will do the Pod->PodTemplateSpec conversion?

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,75 @@
+	badPods := []*v1.Pod{}
+	goodPods := []*v1.Pod{}
+	for _, pod := range pods {
+		if pod.Status.Phase == v1.PodRunning {
```
Terminating pods are still running pods, right?

kargakis replied:
Also we should probably use IsPodReady (uses Conditions) in favor of Phase.

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+
+	glog.V(4).Infof("Marking old pods for deletion")
+	for _, pod := range goodPods {
+		if numUnavailable >= maxUnavailable {
```
Doesn't need to be in the loop.

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,75 @@
+func GetNumUnavailable(ds *extensions.DaemonSet, newPods, oldPods []*v1.Pod) int {
+	unavailable := int(ds.Status.DesiredNumberScheduled) - len(newPods) - len(oldPods)
+	for _, pod := range newPods {
+		if pod.Status.Phase != v1.PodRunning {
```
You need to use IsPodAvailable and provide MinReadySeconds from the ds spec

kargakis replied:
Also pods that are marked for deletion should count as unavailable.

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+	glog.V(4).Infof("Marking all bad old pods for deletion")
+	for _, pod := range badPods {
+		glog.V(4).Infof("Marking pod %s/%s for deletion", ds.Name, pod.Name)
+		podsToDelete = append(podsToDelete, pod.Name)
```
We shouldn't retry the deletion if the pod is already marked for deletion?

lukaszo replied:
I'm not sure. `manage` doesn't check it. Also `RollingUpdate` shouldn't be executed until all expected deletions are complete.

janetkuo replied:
Pods won't be deleted immediately after `Delete` returns; it may be terminating depending on its termination grace period

lukaszo replied:
In `GetNumUnavailable` I'm only counting failed new pods but I see it's misleading now. I will refactor it.

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -716,6 +744,17 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 		}
 	}
 
+	dsNeedsSync = dsc.expectations.SatisfiedExpectations(dsKey)
```
Already called above. Although expectations never worked as expected and we should probably get rid of them from this controller too.

lukaszo replied:
`manage` can set new expectations. I want to run updates only if `manage` didn't change anything.

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+func GetNumUnavailable(ds *extensions.DaemonSet, newPods, oldPods []*v1.Pod) int {
+	unavailable := int(ds.Status.DesiredNumberScheduled) - len(newPods) - len(oldPods)
```
`DesiredNumberScheduled` may not be synced when you call `GetNumUnavailable`. Generally, utilities for constructing the status of an object should not treat fields from the status object as correct. Instead try to identify how much `DesiredNumberScheduled` really is.

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+	if err != nil {
+		return fmt.Errorf("couldn't get list of nodes during rolling update of daemon set %#v: %v", ds, err)
+	}
+	maxUnavailable, err := intstrutil.GetValueFromIntOrPercent(ds.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, len(nodeList), true)
```
There is a helper for this in the deployment utilities.

lukaszo replied:
This one https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L417 ? I can not use it.

janetkuo replied:
This missed the fact that DaemonSets don't always target all nodes (node selector, node affinity... ), see `predicates.GeneralPredicates`

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -523,11 +525,21 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
+	return utilerrors.NewAggregate(errors)
+}
+
+func (dsc *DaemonSetsController) syncNodes(ds *extensions.DaemonSet, podsToDelete, nodesNeedingDaemonPods []string) []error {
```
godoc what syncNodes is supposed to do.

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+	podutil "k8s.io/kubernetes/pkg/controller/deployment/util"
+)
+
+func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
```
godoc what `rollingUpdate` is doing exactly. It seems that it simply deletes pods if possible and then `manage` will make sure the new pods will spin up?

lukaszo replied:
Yes, manage will create new pods but I'm not sure if this is a correct approach. I need to run more tests.

janetkuo replied:
Add a comment to describe that? `// rollingUpdate does such and such...`

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+func (dsc *DaemonSetsController) getAllDaemonSetPods(ds *extensions.DaemonSet) ([]*v1.Pod, []*v1.Pod, error) {
+	newPods := []*v1.Pod{}
```
Use `var newPods []*v1.Pod` instead - consistent with what's the conventional way of defining stuff in the codebase (avoids an extra allocation)

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -659,6 +678,14 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *extensions.DaemonSet)
+					updated = true
+				}
+				if ready && updated {
+					numberAvailable++
```
This is not what available should be. Available is the # of pods (old or new) that are ready for at least minReadySeconds.

janetkuo replied:
Yes, ref https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md#daemonset-controller:

- `.status.numberAvailable` = the total number of DaemonSet pods that have become `Ready` for `MinReadySeconds`
- `.status.numberUnavailable` = `.status.desiredNumberScheduled` - `.status.numberAvailable`

lukaszo replied:
done

idvoretskyi commented:
cc @kubernetes/mirantis

kargakis commented:
Can you paste the errors you are getting? We should be good with generated
conversions for DS afaik

On Wed, Feb 8, 2017 at 4:19 PM, Łukasz Oleś <notifications@github.com>
wrote:

> *@lukaszo* commented on this pull request.
> ------------------------------
>
> In pkg/apis/extensions/v1beta1/conversion.go
> <https://github.com/kubernetes/kubernetes/pull/41116>:
>
> > @@ -219,6 +225,73 @@ func Convert_v1beta1_RollingUpdateDeployment_To_extensions_RollingUpdateDeployme
>  	return nil
>  }
>
> +func Convert_extensions_DaemonSetSpec_To_v1beta1_DaemonSetSpec(in *extensions.DaemonSetSpec, out *DaemonSetSpec, s conversion.Scope) error {
>
> Hmm, I had some errors with validation earlier and thought that it will
> help. It turns out that only
> Convert_extensions_RollingUpdateDaemonSet_To_v1beta1_RollingUpdateDaemonSet
> Convert_v1beta1_RollingUpdateDaemonSet_To_extensions_
> RollingUpdateDaemonSet,
> are required
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41116>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf8XNL153xm7TX44rft-Fa7XnBV0mks5radz-gaJpZM4L6TEV>
> .
>

janetkuo commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -369,27 +369,28 @@ type DeploymentList struct {
-// +optional
-	Type DaemonSetUpdateStrategyType `json:"type,omitempty"`
+	// Type of daemon set update. Can be "RollingUpdate" or "OnDelete".
+	// Default is OnDelete.
```
Add `// +optional` tag to all fields with `omitempty` json tag, and add "optional" tag to relative fields in extensions/types.go too, see #34860 

cc @mbohlool

lukaszo replied:
done

lukaszo replied:
Ah, I didn't add it everywhere.

mbohlool replied:
I didn't read this PR, so ignore my comment if it does not make sense. There are cases that you may want to have `omitempty` but not `+optional`. If your intention is actually making a field optional, then add `+optional` otherwise we can discuss it more.

janetkuo replied:
@mbohlool thanks for the clarification!

janetkuo replied:
@lukaszo most fields in the file still need the `+optional` flag

lukaszo replied:
done

janetkuo commented on pkg/apis/extensions/types.go:
```diff
@@ -482,6 +468,20 @@ type DaemonSetStatus struct {
+	NumberAvailable int32
+
+	// NumberUnavailable is the number of nodes that should be running the
+	// daemon pod and have non of the daemon pod running and available
```
typo: non**e**

lukaszo replied:
done

lukaszo commented:
I pushed new version with the following changes:
- added +optional to fields in types.go
- fixed counting numberAvailable/numberUnavailable
- using minReadySeconds
- switched everywhere to using IsPodReady instead of checking Phase
- optimized code in validation.go

janetkuo commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+	// for oldPods delete all not running pods
+	var podsToDelete []string
+	glog.V(4).Infof("Marking all bad old pods for deletion")
+	for _, pod := range badPods {
```
It should be safe to kill old pods that aren't available yet.

lukaszo replied:
Instead of splitting by running pods I should split by available? Good idea.

janetkuo commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,92 @@
+	for _, pod := range badPods {
+		glog.V(4).Infof("Marking pod %s/%s for deletion", ds.Name, pod.Name)
+		podsToDelete = append(podsToDelete, pod.Name)
+		numUnavailable++
```
...and killing them shouldn't increase `numUnavailable` (they were already unavailable)

lukaszo replied:
In GetNumUnavailable I'm only counting failed new pods but I see it's misleading now. I will refactor it.

janetkuo commented:
Not sure if we still need pod-template-hash for DaemonSets in this version of implementation. It was originally for users to easily filter pods with new/old template. Since now we don't support history yet, the hash number won't make as much sense. Plus we've faced some issues of hash in Deployments, and I'd like to hold on implementing hash for DaemonSets until it's all resolved. WDYT? @lukaszo @kargakis

kargakis commented:
> Not sure if we still need pod-template-hash for DaemonSets in this version of implementation. It was originally for users to easily filter pods with new/old template. Since now we don't support history yet, the hash number won't make as much sense. Plus we've faced some issues of hash in Deployments, and I'd like to hold on implementing hash for DaemonSets until it's all resolved. WDYT? @lukaszo @kargakis

Yes, I agree we don't need hashing. Hashing is needed in Deployments for naming ReplicaSets and since the initial version of DS upgrade won't include history, it seems unnecessary.

lukaszo commented on pkg/api/testing/fuzzer.go:
```diff
@@ -508,6 +508,21 @@ func extensionFuncs(t apitesting.TestingCommon) []interface{} {
+			} else {
+				rollingUpdate := extensions.RollingUpdateDaemonSet{}
+				if c.RandBool() {
+					rollingUpdate.MaxUnavailable = intstr.FromInt(int(c.RandUint64()))
```
done

lukaszo commented:
- refactored GetNumUnavailable to getUnavailableNumbers. Now it returns maxUnavailable and GetNumUnavailable
- maxUnavailable is now correctly calcualtes
- insted of deleting all bad old pods I'm deleting all old unavailable pods
- fixed bug with setting 0 to MaxUnavailable in fuzzer.go
- fixed typo in types.go files

lukaszo commented:
If you don't want to use hashin then what instead of it? Using `apimachinery.Semantic.DeepEqual` to compare  pod.Spec and Ds.Spec.Template will not work. After pod is created its Spec is changed. Some new fields are added for example volume with secrets.
This is Ds.Spec:
`  template:
    metadata:
      creationTimestamp: null
      labels:
        app: datastore
    spec:
      containers:
      - image: gcr.io/google_containers/pause:2.0
        imagePullPolicy: IfNotPresent
        name: datastore
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulername: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30`

And here is a pod.Spec:
`spec:
  containers:
  - image: gcr.io/google_containers/pause:2.0
    imagePullPolicy: IfNotPresent
    name: datastore
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-ddtw9
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: 172.17.0.8
  restartPolicy: Always
  schedulername: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  volumes:
  - name: default-token-ddtw9
    secret:
      defaultMode: 420
      secretName: default-token-ddtw9
`
Or is there another way to compare them?

janetkuo commented:
@lukaszo you're right. Let's keep the hashing

janetkuo commented on pkg/api/testing/fuzzer.go:
```diff
@@ -508,6 +508,21 @@ func extensionFuncs(t apitesting.TestingCommon) []interface{} {
+			} else {
+				rollingUpdate := extensions.RollingUpdateDaemonSet{}
+				if c.RandBool() {
+					rollingUpdate.MaxUnavailable = intstr.FromInt(1 + int(c.RandUint64()))
```
`intstr.FromInt` only takes value no greater than `int32`, so use `int(c.Rand.Int31())` instead. Also, add `else` here for the "%" case.

```go
if c.RandBool() {
	rollingUpdate.MaxUnavailable = intstr.FromInt(1 + int(c.Rand.Int31()))
} else {
	rollingUpdate.MaxUnavailable = intstr.FromString(fmt.Sprintf("%d%%", 1+c.Rand.Int31())
}
```

lukaszo replied:
done

janetkuo commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,66 @@
+
+func GetPodTemplateWithHash(ds *extensions.DaemonSet, dsPodTemplateSpecHash uint32) v1.PodTemplateSpec {
+	obj, _ := api.Scheme.DeepCopy(ds.Spec.Template)
+	dsPodTemplateSpecHashStr := strconv.FormatUint(uint64(dsPodTemplateSpecHash), 10)
```
Suggest just using `fmt.Sprint(dsPodTemplateSpecHash)`, which is simpler

lukaszo replied:
done

janetkuo commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,66 @@
+			return true
+		}
+	} else {
+		// XXX: Hash does not exist. It's old pod. For now returning true.
```
If it's old pod we should return `false` (pod not updated)?

lukaszo replied:
By old I mean it was created in k8s 1.5 and it hasn't a hash.

janetkuo replied:
```go
if !hashExists || curPodTemplateSpecHash == dsPodTemplateSpecHashStr {
  // ...
  return true
}
return false
```

kargakis replied:
How can we be sure that a pod without a hash is updated w/o actually introspecting it?

kargakis replied:
Also about new pods with the hash, afaik these are not immutable so something may change in their spec and the ds controller will continue thinking they are new, right? This is part of a bigger problem, @lukaszo can you open a separate issue about it? I believe we should validate changes in resources owned by other resources but it needs more discussion.

janetkuo commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -548,10 +560,12 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
 	createWait := sync.WaitGroup{}
 	createWait.Add(createDiff)
+	dsPodTemplateSpecHash := podutil.GetPodTemplateSpecHashFnv(ds.Spec.Template)
+	template := util.GetPodTemplateWithHash(ds, dsPodTemplateSpecHash)
```
Suggest putting the "calculating hash" part inside `GetPodTemplateWithHash` and we only pass in `ds.Spec.Template`

kargakis replied:
The reason I broke it down into to separate steps in Deployments is because of the algorithm migration that needs to happen - since this is not needed here, moving the hash calculation in GetPodTemplateWithHash is fine.

lukaszo replied:
done

janetkuo commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -716,6 +742,17 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 		}
 	}
 
+	dsNeedsSync = dsc.expectations.SatisfiedExpectations(dsKey)
```
It's not part of your change, but we don't need to get `dsKey` from `KeyFunc`, right? We could just use the input `key`

janetkuo commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,66 @@
+
+func IsPodUpdated(dsPodTemplateSpecHash uint32, pod *v1.Pod) bool {
+	curPodTemplateSpecHash, hashExists := pod.ObjectMeta.Labels[extensions.DefaultDaemonSetUniqueLabelKey]
+	dsPodTemplateSpecHashStr := strconv.FormatUint(uint64(dsPodTemplateSpecHash), 10)
```
`fmt.Sprint(dsPodTemplateSpecHash)` here as well

janetkuo replied:
Also suggest passing in ds template and calculate hash in `IsPodUpdated`, so that you don't need to calculate hash and pass it in every time you call `IsPodUpdated`. Also all public functions need godoc, a comment describing what this function does

lukaszo replied:
If I pass ds template here I will need to calculate hash every time this function is called. Now I calculate in once before entering the loop where IsPodUpdated is called.

kargakis replied:
> If I pass ds template here I will need to calculate hash every time this function is called. Now I calculate it once before entering the loop where IsPodUpdated is called.

Sounds reasonable

janetkuo commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,120 @@
+		}
+		if wantToRun {
+			desiredNumberScheduled++
+		}
```
We should use `dsc.getNodesToDaemonPods(ds)` and calculate unavailable pod on each node. If there are more than one daemon pod on a particular node, we should only calculate their unavailability once (at most one available/unavailable daemon pod per node per DS).

kargakis replied:
Agreed

lukaszo replied:
I'm running it after manage so there should be no more than one pod per node. Your way seems more consistent though. I will change it.

lukaszo replied:
done

kargakis commented:
The problem with hashing the pod template is that every time the pod template api changes, the hash will change (see https://github.com/kubernetes/kubernetes/issues/40415). How about using a new PodTemplateGeneration field that will be incremented in a similar fashion to metadata.Generation but only in PodTemplate updates and then hash that field? It should always produce a stable hash.

@smarterclayton @mfojtik @kubernetes/sig-apps-misc thoughts?

kargakis commented:
> If you don't want to use hashing then what instead of it? Using apimachinery.Semantic.DeepEqual to compare pod.Spec and Ds.Spec.Template will not work. After pod is created its Spec is changed. Some new fields are added for example volume with secrets.

I had envisioned an api helper that would return you the PodTemplateSpec given a Pod.

kargakis commented on pkg/apis/extensions/types.go:
```diff
@@ -482,6 +471,23 @@ type DaemonSetStatus struct {
+	NumberAvailable int32
+
+	// NumberUnavailable is the number of nodes that should be running the
+	// daemon pod and have none of the daemon pod running and available
```
This is not clear to me. Does it include nodes where the pod does not exist at all or only nodes where the pod is created but not yet ready for minReadySeconds? In Deployment, UnavailableReplicas maps to the latter.

kargakis replied:
Although I think I would be fine with this meaning the former for DaemonSets.

lukaszo replied:
If the pod is missing I'm counting it as unavailable so the first option.

janetkuo replied:
This also implies that if a node has 2 daemon pods (from the same DS), where one pod is available but the other one is not, we won't count this as unavailable. 

Should we rephrase it to make it more clear? How about this:

```go
// NumberUnavailable is the number of nodes that should be running the
// daemon pod but have no available daemon pod
```

krmayankk replied:
@janetkuo confused how come a node can have 2 daemon pods from the same DS. The DS has one podspec and can only schedule one pod per node for a given DS spec.

janetkuo replied:
@krmayankk We use cache in controllers (e.g. while listing pods), so it's theoretically possible that a DaemonSet controller create more than one DaemonSet pod on a node transiently.

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,120 @@
+	// for oldPods delete all not running pods
+	var podsToDelete []string
+	glog.V(4).Infof("Marking all unavailable old pods for deletion")
+	for _, pod := range oldUnavailablePods {
```
Won't you end up deleting pods before they actually get a chance to become available? For example, a DaemonSet with MinReadySeconds=60 will spin up a new Pod, the DS will be resynced and in the next iteration which most likely will be sooner than 60s, the new Pod will be deleted.

lukaszo replied:
I'm deleting only old pods. When new pod is created(with new hash) I will not touch it.

janetkuo replied:
Yes, only _old_ unavailable pods are deleted here. @kargakis are you suggesting it's safer to just kill old pods that are not running (i.e. don't kill the running but unavailable old pods)? Or we don't care since it's old?

kargakis replied:
If we can kill it either because it's allowed by our budget or is a duplicate, we won't care.

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func (dsc *DaemonSetsController) GetUnavailableNumbers(ds *extensions.DaemonSet, allPods []*v1.Pod) (int, int, error) {
+	nodeList, err := dsc.nodeLister.List(labels.Everything())
```
A single NodeList can be consolidated at the start of the worker function that processes a DaemonSet and passed around to whichever function that needs it. A TODO would suffice for now.

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,66 @@
+	unavailablePods := []*v1.Pod{}
+	availablePods := []*v1.Pod{}
+	for _, pod := range pods {
+		if v1.IsPodAvailable(pod, ds.Spec.MinReadySeconds, metav1.Now()) {
```
TODO to switch the DS controller to use a virtual clock otherwise MinReadySeconds is not reliable.

lukaszo replied:
What's virtual clock?

kargakis replied:
Not sure of the exact library, I would expect https://github.com/kubernetes/client-go/blob/master/util/clock/clock.go but I haven't set up one yet. @gmarek @smarterclayton will know better of the particulars, supposedly the node controller uses one. The gist is that it's needed by controllers that interact with pods so they can record their own times for things like transitions to ready state instead of relying in times set by the kubelet (node and master clocks may skew: https://github.com/kubernetes/kubernetes/issues/29229)

kargakis commented:
Is MinReadySeconds really necessary for including in the initial implementation? If so, you will need to something similar to what https://github.com/kubernetes/kubernetes/pull/38076 is doing for ReplicaSets.

lukaszo commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -657,18 +675,26 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *extensions.DaemonSet)
 				// Sort the daemon pods by creation time, so the the oldest is first.
```
done

lukaszo commented:
I pushed new changes
- changed fuzzer.go to use c.Rand.Int31() and also to generate version with percentage
- added +optional to new fields in types.go files
- added documentation
- moved calculating hash to GetPodTemplateWithHash
- fixed typo in comment
- changed getNumUnavailable to use dsc.getNodesToDaemonPods 
- changed rollingUpdate to not delete pods which are already marked for deletion

janetkuo commented:
> How about using a new PodTemplateGeneration field that will be incremented in a similar fashion to metadata.Generation but only in PodTemplate updates and then hash that field? It should always produce a stable hash.

@kargakis would you explain more on how this works? Do we also need something like PodTemplateObservedGeneration? Where does this field live (in DaemonSet?) and when does it get updated?

janetkuo commented:
> I had envisioned an api helper that would return you the PodTemplateSpec given a Pod.

I assume this helper would use pod default values to infer the PodTemplateSpec used to create that pod? However, pod default values may change across k8s versions. How does the helper know the PodTemplateSpec given a Pod?

janetkuo commented on pkg/api/testing/fuzzer.go:
```diff
@@ -508,6 +508,25 @@ func extensionFuncs(t apitesting.TestingCommon) []interface{} {
+				j.RollingUpdate = nil
+			} else {
+				rollingUpdate := extensions.RollingUpdateDaemonSet{}
+				if c.RandBool() {
```
We don't need this `c.RandBool()`, since we'll always set `rollingUpdate.MaxUnavailable` to something greater than 0 on RollingUpdate strategy

lukaszo replied:
This bool says if we should set it. It's optional so it doesn't need to be set.

janetkuo commented on pkg/apis/extensions/types.go:
```diff
@@ -436,29 +431,23 @@ type DaemonSetSpec struct {
-		// DefaultDaemonSetUniqueLabelKey("daemonset.kubernetes.io/podTemplateHash").
-		// Value of this key is hash of DaemonSetSpec.PodTemplateSpec.
-		// No label is added if this is set to empty string.
+	// MinReadySeconds minimum number of seconds for which a newly created DaemonSet pod should
```
`MinReadySeoncds is...` or just remove MinReadySeconds (`Minimum number of...`)

janetkuo commented on pkg/apis/extensions/types.go:
```diff
@@ -482,6 +471,23 @@ type DaemonSetStatus struct {
+	UpdatedNumberScheduled int32
+
+	// NumberAvailable is the number of nodes that should be running the
+	// daemon pod and have one or more of the daemon pod running and
```
Maybe remove "running" since a ready/available pod must be running? Just trying to make it more understandable:

```go
// NumberAvailable is the number of nodes that should be running the
// daemon pod and have at least one available daemon pod (ready for
// at least minReadySeconds)
```

kargakis commented:
> @kargakis would you explain more on how this works? Do we also need something like PodTemplateObservedGeneration? Where does this field live (in DaemonSet?) and when does it get updated?

A field in the spec that gets updated every time the podtemplatespec changes, running as a rest hook. I don't think this as similar to the revision of the Deployment, that's an annotation maintained by the controller which may get out of sync and potentially lost because it depends on your available history. 

Then, hash *name*-podtemplategen.
When we end up with duplicate replica sets we can migrate the old to the new as a normal rollout or delete in case it has no replicas.

We don't need a separate observed generation for this one.

kargakis commented:
> I assume this helper would use pod default values to infer the PodTemplateSpec used to create that pod? However, pod default values may change across k8s versions. How does the helper know the PodTemplateSpec given a Pod?

Pods created from templates owned by controllers should be immutable apart from a couple of fields imo. Changing defaults is a breaking change but I am not sure that we need to do something about it since we only care if the current podtemplatespec that can be inferred from a pod equals the ds podtemplate (and both should be affected by new defaults).

krmayankk commented on api/openapi-spec/swagger.json:
```diff
@@ -41103,6 +41112,11 @@
       "format": "int32"
      },
+     "numberAvailable": {
+      "description": "NumberAvailable is the number of nodes that should be running the daemon pod and have one or more of the daemon pod running and available (ready for at least minReadySeconds)",
```
does this include both current and updated pods ?

kargakis replied:
Yes.

krmayankk commented on pkg/apis/extensions/v1beta1/defaults.go:
```diff
@@ -47,6 +47,21 @@ func SetDefaults_DaemonSet(obj *DaemonSet) {
+		}
+		if updateStrategy.RollingUpdate.MaxUnavailable == nil {
+			// Set default MaxUnavailable as 1 by default.
+			maxUnavailable := intstr.FromInt(1)
```
i am guessing this default of 1 is intentional, since i have seen a proposal where we agreed on 25% for Deployments.

kargakis replied:
Deployment has different requirements for availability and it can also have surge pods (more than desired).

kargakis commented:
I have updated my latest comments regarding hashing and creating PodTemplateSpecs out of pods. I really believe we shouldn't use hashing for DaemonSets before we end up with a stable solution for Deployments.

I don't think given a pod that it would be difficult to infer a PodTemplateSpec, and then compare that PodTemplateSpec with what's in the DaemonSet in order to determine if the running pod is new or old.

lukaszo commented:
Ok, I will try to implement it.

janetkuo commented:
> The basic idea is:
> 1. Hash {metadata.name of Deployment/DaemonSet}-podtemplategen. 
> 2. Use the hash to construct the name of the {ReplicaSet for Deployments, PodTemplate or whatever we use for DaemonSets}.

Why do we need to hash? To prevent name collisions?

> 3. When we end up with duplicate {ReplicaSets,PodTemplates} we can migrate the old to the new as a normal rollout or delete the old in case it has no underlying replicas.

What do you mean by duplicate? Did you mean the rollback case?

@kargakis IIUC, this will depend on the to-be-implemented helper function that gets pod template from pod spec, to observe template spec changes?

janetkuo commented:
> Pods created from templates owned by controllers should be immutable apart from a couple of fields imo. Changing defaults is a breaking change but I am not sure that we need to do something about it since we only care if the current podtemplatespec that can be inferred from a pod equals the ds podtemplate (and both should be affected by new defaults).

@kargakis I've discussed with people working on API machinery and kubelet. I think it'd be hard to implement such a function. There are many default values being applied to a pod upon creation (e.g. nodeName, ...), and admission controllers may also modify pods.

kargakis commented:
> Why do we need to hash? To prevent name collisions?

For Deployments, this entails we continue to use the same naming scheme as we already do. Also the name does not imply a sequence. foo-1 is expected to be older and likely replaced by foo-2.

> What do you mean by duplicate? Did you mean the rollback case?

Duplicates have the same podtemplatespec.

kargakis commented:
> @kargakis I've discussed with people working on API machinery and kubelet. I think it'd be hard to implement such a function. There are many default values being applied to a pod upon creation (e.g. nodeName, ...), and admission controllers may also modify pods.

Node names are usually not specified in templates but I can see this being a problem with admission controllers. What values are currently updated?

lukaszo commented:
New changes:
- added two e2e tests
- rebased generated part

janetkuo commented:
> Node names are usually not specified in templates but I can see this being a problem with admission controllers. What values are currently updated?

For example, image pull policy.

janetkuo commented:
I thought about it and had discussions with different sigs, and I think `PodTemplateGeneration` is the right way to go. 

1. In DaemonSet update strategy, PodTemplateGeneration++ on DS.spec.template updates
2. When DS creates pods, label them with current PodTemplateGeneration ("pod-template-generation=n")
    - Use labels since it's easy for the controller and the users to filter pods that match current podtemplate generation.
3. DS compares pods' "pod-template-generation" label against its own PodTemplateGeneration to decide to kill (update) this pod or not.

One drawback is that it's not extensible (customized controllers that extend DaemonSet controller will have to modify API server code as well) - we decided not to worry about this for now. 

Another drawback is that in the rollback case (in the future), we may restart more pods than we want (since we can only roll forward PodTemplateGeneration), and pods gen 1 may actually be the same as DS gen 3. 

We can solve the rollback case with DaemonSet history + pod template hash. We use history objects (maybe stand-alone PodTemplate objects?) to store pod-template-hash and templates, and we compare pods' pod-template-hash labels with history and compare history template with DaemonSet template. 

**Alternatives:** 
1. I thought about using `json.Marshal` before hashing the object so that new fields with `omitempty` wouldn't change the hash value. However, if a new fields was added to the PodTemplateAPI with some default value, the hash will still change. 
2. White-listing fields we care about when hashing pod template... this is not extensible. 
3. Implement a helper function that gives you podtemplate given a pod spec. This is hard to implement, as many default values will be added to a pod, and admission controller will also update the pod.

lukaszo commented:
Using history objects is what I had in previous PR so I like it :) @smarterclayton instead of using api.PodTemplate proposed to create a separate new type. 

As for PodTemplateGeneration if we are going to implement it then when and where should it be updated? Somewhere in the api server? I'm wondering if it's a dead end. In a feature we will need history object and PodTemplateGeneration will be unneeded.

janetkuo commented:
> Using history objects is what I had in previous PR so I like it :) @smarterclayton instead of using api.PodTemplate proposed to create a separate new type.

Yes, and we will implement history post-1.6

> As for PodTemplateGeneration if we are going to implement it then when and where should it be updated? Somewhere in the api server? 

In DaemonSet update strategy https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/daemonset/strategy.go#L58-L80
Similar to Generation, we check if the template is changed, and PodTemplateGeneration +1 if so 

> I'm wondering if it's a dead end. In a feature we will need history object and PodTemplateGeneration will be unneeded.

It's one more check against pod template updates: we first check if the pod template generation matches, if not then check pod template hash , if not then deep equal the template (the later two need history being implemented.) It's still forward compatible.

lukaszo commented:
Ah, I didn't know it's that easy. I will prepare the patch. It shouldn't take much time. 
It's api change, do we need to update proposal and get acceptance for it?

janetkuo commented:
> It's api change, do we need to update proposal and get acceptance for it?

Sure, I'll update the proposal accordingly.

lukaszo commented:
I have some questions. Generation is stored in ObjectMeta and it's read only. I can not put PodTemplateGeneration there as it's DaemonSet specific so I should I put in DaemonSet object? 

Another question. Is there some magic way to  make it read only? Or do I need to prevent changes manually?

janetkuo commented:
> I have some questions. Generation is stored in ObjectMeta and it's read only. I can not put PodTemplateGeneration there as it's DaemonSet specific so I should I put in DaemonSet object?

I agree that we should put it in `DaemonSet`.

> Another question. Is there some magic way to make it read only? Or do I need to prevent changes manually?

Generation update is prevented in rest/update.go#BeforeUpdate by always keeping the old Generation
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/update.go#L78-L119:

```go
objectMeta.Generation = oldMeta.Generation
```

It's placed there because Generation is general. We can make PodTemplateGeneration read-only by placing similar logic in `PrepareForUpdate` (we first copy old PodTemplateGeneration to the new one, and increase it by one if there's any template change).

Also I'm thinking naming it TemplateGeneration instead of PodTemplateGeneration, since it's the generation of DaemonSetSpec.Template, not DaemonSetSpec.PodTemplate, WDYT?

janetkuo commented:
PR for updating the proposal https://github.com/kubernetes/community/pull/366

smarterclayton commented:
You would implement that in the Strategy type for daemonsets.

On Feb 14, 2017, at 6:39 PM, Łukasz Oleś <notifications@github.com> wrote:

I have some questions. Generation is stored in ObjectMeta and it's read
only. I can not put PodTemplateGeneration there as it's DaemonSet specific
so I should I put in DaemonSet object?

Another question. Is there some magic way to make it read only? Or do I
need to prevent changes manually?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/41116#issuecomment-279872142>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p-NTr01bYF-128e2nsL0Izd8b8UBks5rcjs8gaJpZM4L6TEV>
.

kargakis commented:
> One drawback is that it's not extensible (customized controllers that extend DaemonSet controller will have to modify API server code as well) - we decided not to worry about this for now.

Can you explain why? I still think this is extensible as long as the DS controller allows it.

kargakis commented:
> Another drawback is that in the rollback case (in the future), we may restart more pods than we want (since we can only roll forward PodTemplateGeneration), and pods gen 1 may actually be the same as DS gen 3.

Not sure I understand this either. We kill pods so there is no restarting involved, right? If we want to rollback from 2 to 1, we spin up any missing pods and relabel the remaining from 1 to 3.

lukaszo commented:
New changes:
- added new TemplateGeneration field in DaemonSet struct according to the proposal https://github.com/kubernetes/community/pull/366
- I've changed IsPodUpdated function tu use TemplateGeneration instead of hashes

There is still open question how to handle old(pre 1.6) daemon pods which don't have a label with generation. I'm experimenting with comparing pod.Spec and daemon.Spec.Template.

lukaszo commented:
@k8s-bot cvm gke e2e test this

kargakis commented on pkg/registry/extensions/daemonset/strategy.go:
```diff
@@ -77,6 +78,14 @@ func (daemonSetStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, ol
+	newDaemonSet.TemplateGeneration = oldDaemonSet.TemplateGeneration
+
+	// Any changes to the Spec.Template increment the TemplateGeneration number.
+	if !reflect.DeepEqual(oldDaemonSet.Spec.Template, newDaemonSet.Spec.Template) {
```
Move this before the metadata generation check, increment both generations and return so we will not run deepEqual twice when not needed.

lukaszo replied:
Move it before? I don't understand? I may move inside the previous `if` clause to be more optimal.

kargakis replied:
```go
if !reflect.DeepEqual(oldDaemonSet.Spec.Template, newDaemonSet.Spec.Template) {
    newDaemonSet.TemplateGeneration = oldDaemonSet.TemplateGeneration + 1
    newDaemonSet.Generation = oldDaemonSet.Generation + 1
    return
}
if !reflect.DeepEqual(oldDaemonSet.Spec, newDaemonSet.Spec) {
    newDaemonSet.Generation = oldDaemonSet.Generation + 1
}
```

lukaszo replied:
Hmm, it looks overcomplicated but I don't have strong opinion here. I will change it

lukaszo replied:
done

kargakis commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -110,6 +110,9 @@ func validateDaemonSetStatus(status *extensions.DaemonSetStatus, fldPath *field.
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberAvailable), fldPath.Child("numberAvailable"))...)
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(status.NumberUnavailable), fldPath.Child("numberUnavailable"))...)
 	return allErrs
 }
```
ValidateDaemonSetUpdate needs to validate that templateGeneration is never decremented or incremented more than one.

lukaszo replied:
Is it needed? I've made it read only.

janetkuo replied:
Agree that this should never happen, but it wouldn't hurt to add one more validation

kargakis commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -141,6 +144,39 @@ func ValidateDaemonSetSpec(spec *extensions.DaemonSetSpec, fldPath *field.Path)
 	if spec.Template.Spec.RestartPolicy != api.RestartPolicyAlways {
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("template", "spec", "restartPolicy"), spec.Template.Spec.RestartPolicy, []string{string(api.RestartPolicyAlways)}))
 	}
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(spec.MinReadySeconds), fldPath.Child("minReadySeconds"))...)
```
You also need to validate that templateGeneration is non-negative.

lukaszo replied:
The same as above. I've made it read only.

gmarek commented:
cc @piosz

lukaszo committed:
DaemonSet updates

It implements https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md
Feature https://github.com/kubernetes/features/issues/124

lukaszo commented:
@k8s-bot gci gce e2e test this

lukaszo commented:
New changes:

* renamed util.GetPodTemplateWithRevision to util.GetPodTemplateWithGeneration
* added more unit tests

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -652,21 +668,29 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *extensions.DaemonSet)
+				if util.IsPodUpdated(ds.TemplateGeneration, pod) {
+					updatedNumberScheduled++
+				}
+				if v1.IsPodAvailable(pod, ds.Spec.MinReadySeconds, metav1.Now()) {
```
Move this inside the `if v1.IsPodReady(pod) {` clause

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,138 @@
+
+	maxUnavailable, numUnavailable, err := dsc.getUnavailableNumbers(ds, allPods)
+	if err != nil {
+		glog.Errorf("Couldn't get unavailable numbers: %v", err)
```
All glog.Errorf calls should switch to utilruntime.HandleError

lukaszo replied:
Actually I should return error here, instead of returning the nil.

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,138 @@
+		if err != nil {
+			return -1, -1, err
+		}
+		if wantToRun {
```
style:
```go
if !wantToRun {
   continue
}
```
and de-indent the rest of the code

lukaszo replied:
done

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,138 @@
+				continue
+			}
+			available := false
+			for _, pod := range daemonPods {
```
We shouldn't have mutliple pods here - make sure you don't count something as available and then delete it as a duplicate.

lukaszo replied:
Yes, this list should have only one element at this point. 
To 200% sure I can sort it and check only the oldest pod. The same way as I did in status.

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,70 @@
+// GetPodTemplateWithHash returns copy of provided template with additional
+// label which contains hash of provided template
+func GetPodTemplateWithGeneration(template v1.PodTemplateSpec, generation int64) v1.PodTemplateSpec {
+	obj, _ := api.Scheme.DeepCopy(template)
```
@janetkuo can you verify that we need this in the Deployment helpers too?

janetkuo replied:
Yeah, we should probably do this for Deployment helper

kargakis commented on pkg/controller/daemon/util/daemonset_util.go:
```diff
@@ -0,0 +1,70 @@
+		}
+	} else {
+		// XXX: Hash does not exist. It's old pod. For now returning true.
+		return true
```
If we are going to kick-off rollouts as soon as a user switches to RollingUpdate, this should be false.
A more concise function:
```go
podTemplateGeneration, generationExists := pod.ObjectMeta.Labels[extensions.DaemonSetTemplateGenerationKey]
dsTemplateGenerationStr := fmt.Sprint(dsTemplateGeneration)
return generationExists && podTemplateGeneration == dsTemplateGenerationStr
```

lukaszo replied:
done

kargakis commented:
> you will need to something similar to what #38076 is doing for ReplicaSets.

@lukaszo fyi, this is still pending. Don't forget, otherwise MinReadySeconds is totally broken.

lukaszo commented:
> @lukaszo fyi, this is still pending. Don't forget, otherwise MinReadySeconds is totally broken.

I've forgot :( Will do

lukaszo commented:
New changes:
- rebased
- fixes after @kargakis comments

lukaszo commented:
@k8s-bot gci gce etcd3 e2e test this

lukaszo commented:
Implemented changes from #38076 (basically copy/paste of the code)

lukaszo commented:
@k8s-bot gci gce kops aws non-cri etcd3 unit cvm gce e2e  test this

kargakis commented:
@lukaszo can you add incremental commits to enable easier reviewing? It's really difficult to re-review the whole thing. You can squash after the PR is lgtm'ed.

kargakis commented:
> @lukaszo can you add incremental commits to enable easier reviewing? It's really difficult to re-review the whole thing. You can squash after the PR is lgtm'ed.

For any of your upcoming changes.

lukaszo commented:
> @lukaszo can you add incremental commits to enable easier reviewing? It's really difficult to re-review the whole thing. You can squash after the PR is lgtm'ed.

Sure, will do

janetkuo commented:
The proposal for using `templateGeneration` is now merged https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md

Note that we bump `templateGeneration` in update strategy. We also allow users to set `templateGeneration` when (and only when) they create the DaemonSet, so that they can adopt existing DaemonSet pods without restarting them. 

We don't use hash as discussed.

janetkuo commented:
Note that 1.6 code freeze starts at Monday, February 27, 2017 @ 6PM PST. We need to get this LGTM'ed before that.

janetkuo commented on pkg/apis/extensions/types.go:
```diff
@@ -506,8 +505,20 @@ type DaemonSet struct {
+	// A sequence number representing a specific generation of the template.
+	// Populated by the system. Read-only.
+	// +optional
+	TemplateGeneration int64
```
Let's move it to spec. We also need to update its comment to say that we allow users to set this value when creating a DaemonSet, see [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md)

janetkuo commented on pkg/registry/extensions/daemonset/strategy.go:
```diff
@@ -53,6 +53,7 @@ func (daemonSetStrategy) PrepareForCreate(ctx genericapirequest.Context, obj run
 	daemonSet.Status = extensions.DaemonSetStatus{}
 
 	daemonSet.Generation = 1
+	daemonSet.TemplateGeneration = 1
```
We allow users to set this value on creation time, so only set this to 1 if it's unset (i.e. any value that's <=0).

lukaszo committed:
Implement kubectl rollout status for daemon sets

lukaszo committed:
Move TemplateGeneration to the Spec

lukaszo committed:
Generated part for DaemonSet updates

lukaszo commented:
New changes:

- moved TemplateGeneration to the Spec
- implemented kubectl rollout status for daemonsets

I also removed [WiP] from the title. I believe that everything is implemented now.

lukaszo commented:
@k8s-bot gci gke e2e test this

lukaszo commented:
@k8s-bot unit test this

janetkuo commented on pkg/controller/daemon/update_test.go:
```diff
@@ -0,0 +1,35 @@
+	syncAndValidateDaemonSets(t, manager, ds, podControl, 5, 0)
+	syncAndValidateDaemonSets(t, manager, ds, podControl, 5, 0)
+	// change strategy type to RollingUpdate
+	ds.Spec.UpdateStrategy.Type = extensions.RollingUpdateDaemonSetStrategyType
```
This test looks unfinished?

lukaszo replied:
I will remove it and move here update tests from daemonsetcontroller_test.go

janetkuo replied:
sounds good

janetkuo commented on pkg/controller/daemon/util/daemonset_util_test.go:
```diff
@@ -0,0 +1,87 @@
+		},
+		{
+			int64(12355),
+			newPod("pod1", "node1", map[string]string{}),
```
Suggest adding a test for the case that label==nil

lukaszo replied:
done

janetkuo commented on pkg/controller/daemon/util/daemonset_util_test.go:
```diff
@@ -0,0 +1,87 @@
+	newPodTemplate := GetPodTemplateWithGeneration(podTemplateSpec, generation)
+	label, exists := newPodTemplate.ObjectMeta.Labels[extensions.DaemonSetTemplateGenerationKey]
+	if !exists || label != fmt.Sprint(generation) {
+		t.Errorf("Error in getting podTemplateSpec with label geneartion. Exists: %t, label: %s", exists, label)
```
typo: generation

lukaszo replied:
done

janetkuo commented on pkg/controller/daemon/util/daemonset_util_test.go:
```diff
@@ -0,0 +1,87 @@
+	}
+}
+
+func TestGetPodTemplateWithRevision(t *testing.T) {
```
`TestGetPodTemplateWithGeneration`

lukaszo replied:
done

janetkuo commented on pkg/kubectl/rollout_status.go:
```diff
@@ -77,3 +83,23 @@ func (s *DeploymentStatusViewer) Status(namespace, name string, revision int64)
+
+// Status returns a message describing daemon set status, and a bool value indicating if the status is considered done
+func (s *DaemonSetStatusViewer) Status(namespace, name string, revision int64) (string, bool, error) {
+	//ignoring revision as DaemonSets do not have history yet
```
Handle the case when update strategy isn't `RollingUpdate` - just say that only RollingUpdate is supported

lukaszo replied:
There is no such check for Deployment. Should we fix it there also(in another PR)?

janetkuo replied:
Wait, I was wrong. We support both DaemonSet update strategy (and Deployment as well).

lukaszo replied:
Ah, indeed.

janetkuo reviewed (approved):
Looks good overall, just some nits/suggestions

lukaszo committed:
Moved tests and fixed typs

janetkuo commented:
/approve

janetkuo commented:
/lgtm

janetkuo commented:
@lukaszo I added a release note to the PR description. Feel free to rephrase it if you want to.

lukaszo commented:
@kargakis @smarterclayton can you take a look and approve?

lukaszo commented:
@janetkuoDo I need approval from someone from the list https://github.com/kubernetes/kubernetes/blob/master/OWNERS or can you do it manually?

All api changes were already accepted in the proposal.

janetkuo commented:
Manually approving

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -255,6 +256,17 @@ func (dsc *DaemonSetsController) enqueueDaemonSet(ds *extensions.DaemonSet) {
+		return
+	}
+
+	// TODO: Handle overlapping controllers better. See comment in ReplicationManager.
```
We are going to handle overlapping daemonsets once we make them user owner references.

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -521,11 +539,23 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
+	return utilerrors.NewAggregate(errors)
+}
+
+// syncNodes deletes given pods and creates new daemon set pods on the given node
```
seems to work per ds and not per node

lukaszo replied:
ah, it should be `on the given nodes`. I will fix it after merge :)

kargakis commented on pkg/controller/deployment/util/pod_util.go:
```diff
@@ -44,6 +44,8 @@ func GetInternalPodTemplateSpecHash(template api.PodTemplateSpec) uint32 {
 	return podTemplateSpecHasher.Sum32()
 }
 
+// TODO: move it to a better place. It's also used by DaemonSets
```
This comment is no longer true

kargakis commented on pkg/kubectl/rollout_status.go:
```diff
@@ -77,3 +83,23 @@ func (s *DeploymentStatusViewer) Status(namespace, name string, revision int64)
+		if daemon.Status.UpdatedNumberScheduled < daemon.Status.DesiredNumberScheduled {
+			return fmt.Sprintf("Waiting for rollout to finish: %d out of %d new pod have been updated...\n", daemon.Status.UpdatedNumberScheduled, daemon.Status.DesiredNumberScheduled), false, nil
+		}
+		if daemon.Status.NumberAvailable < daemon.Status.DesiredNumberScheduled {
```
This should be `ds.Status.NumberAvailable < ds.Status.DesiredNumberScheduled-MaxUnavailable`

lukaszo replied:
Maybe it should be somehow configurable. I'm doing it in the same way as in status for deployments.

kargakis replied:
I have a PR that updates it for Deployments
https://github.com/kubernetes/kubernetes/pull/41809

kargakis commented on test/e2e/daemon_set.go:
```diff
@@ -231,6 +232,64 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
+		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkRunningOnAllNodes(f, label))
+		Expect(err).NotTo(HaveOccurred(), "error waiting for daemon pod to start")
+	})
+
```
We need a test that verifies adoption

kargakis replied:
And another test that verifies ds availability

lukaszo replied:
I will add them in a new PR

kargakis commented:
We need more e2e tests and kubectl rollout status should probably respect strategy params and quit as soon as the minimum available pods are updated. Fine with a follow-up

lgtm

lukaszo commented:
@k8s-bot non-cri e2e test this

lukaszo commented:
@k8s-bot gce etcd3 e2e test this

lukaszo commented:
Nothing changed, just rebased and updated generated part.

kargakis commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -508,8 +507,20 @@ type DaemonSet struct {
 	Status DaemonSetStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+
+	// A sequence number representing a specific generation of the template.
+	// Populated by the system. Read-only.
```
This is not entirely true since users can populate it at creation time, right?

lukaszo replied:
I will prepare follow up PR with a fix

kargakis commented on pkg/controller/daemon/update.go:
```diff
@@ -0,0 +1,137 @@
+		}
+		available := false
+		for _, pod := range daemonPods {
+			if v1.IsPodAvailable(pod, ds.Spec.MinReadySeconds, metav1.Now()) {
```
getNodesToDaemonPods does not filter out terminating pods and I don't think IsPodAvailable filter them out either. You should consider them as unavailable so before this check you probably want to continue on deletiontimestamp != nil.

kargakis commented:
@k8s-bot unit test this issue https://github.com/kubernetes/kubernetes/issues/39471

kargakis commented:
P2 because it has been rebased a couple of times already and needs to land sooner than more of the PRs in the queue.

lukaszo commented:
@k8s-bot unit test this
